test/unit/StakingProducts/recalculateEffectiveWeight.js

```diff
@@ -0,0 +1,447 @@
+    const productId = Zero;
+
+    await setStakedProducts.call(this, { productIds: [productId], targetWeight: 10 });
+
+    // deposit stake
+    await depositTo.call(this, { staker, amount: stakeAmount });
+
+    // buy cover
+    await allocateCapacity.call(this, { coverBuyer, amount: coverBuyAmount, productId });
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productId]);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(10);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(10);
+
+    // lower target weight to 1
+    await setStakedProducts.call(this, {
+      productIds: [productId],
+      targetWeight: 1,
+    });
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(1);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(8);
+
+    // expire cover. effective weight should be reduced to target weight
+    await increaseTime(daysToSeconds(365));
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productId]);
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(1);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(1);
+    {
+      const stakedProduct = await stakingProducts.getProduct(this.poolId, productId);
+      expect(stakedProduct.lastEffectiveWeight).to.be.equal(1);
+      expect(stakedProduct.targetWeight).to.be.equal(1);
+    }
+  });
+
+  it('effective weight should be lowered from extra deposits', async function () {
+    const { stakingProducts } = this;
+    const [staker, coverBuyer] = this.accounts.members;
+
+    const amount = parseEther('10000');
+    // buy a quarter of the capacity
+    const coverBuyAmount = amount.mul(25).div(100).mul(2); // 8% of capacity with 2x capacity multiplier
+
+    const productIdToAdd = Zero;
+    await setStakedProducts.call(this, { productIds: [productIdToAdd], targetWeight: 25 });
+
+    // deposit stake
+    await depositTo.call(this, { staker, amount });
+
+    // buy cover
+    await allocateCapacity.call(this, { coverBuyer, amount: coverBuyAmount, productId: productIdToAdd });
+
+    // lower target weight to 0
+    await setStakedProducts.call(this, { productIds: [productIdToAdd], targetWeight: 0 });
+
+    expect(await stakingProducts.getTotalTargetWeight(this.poolId)).to.be.equal(0);
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(25);
+
+    // double stake
+    await depositTo.call(this, { staker, amount });
+
+    expect(await stakingProducts.getTotalEffectiveWeight(this.poolId)).to.be.equal(25);
+
+    // recalculate effective weight
+    await stakingProducts.recalculateEffectiveWeights(this.poolId, [productIdToAdd]);
+
+    const expectedEffectiveWeight = One.mul(25).div(2);
```

Comment:
again, ideally this is computed as much as possible

and named constants instead of repeating the same number many times (in this case the 25 for example